solutions/cfgym-101102D.mdx

```diff
@@ -0,0 +1,258 @@
+		for (int r = 0; r < row_num; r++) {
+			for (int c = 0; c < col_num; c++) {
+				std::cin >> grid[r][c];
+			}
+		}
+
+		vector<vector<int>> longest_consec(row_num, vector<int>(col_num));
+		for (int r = 0; r < row_num; r++) {
+			int curr = -1;
+			int streak = 0;
+			for (int c = 0; c < col_num; c++) {
+				if (grid[r][c] == curr) {
+					streak++;
+				} else {
+					curr = grid[r][c];
+					streak = 1;
+				}
+				longest_consec[r][c] = streak;
+			}
+		}
+
+		long long total = 0;
+		for (int c = 0; c < col_num; c++) {
+			vector<pair<int, int>> rel_streaks;
+			int curr_sum = 0;  // the current amt of valid submatrices
+			int curr = grid[row_num - 1][c];
+			// for each row in reverse, count # of valid submatrices that start at that cell
+			for (int r = row_num - 1; r >= 0; r--) {
+				// if we encounter a diff number, just up & reset everything
+				if (grid[r][c] != curr) {
+					curr = grid[r][c];
+					curr_sum = 0;
+					rel_streaks = vector<pair<int, int>>();
+				}
+
+				/*
+				 * the total # of elements that were killed
+				 * between this element & the last one
+				 */
+				int popped = 0;
+				// while this current row can possibly be a chokehold
+				while (!rel_streaks.empty()
+						&& longest_consec[r][c] < rel_streaks.back().first) {
+					// subtract the # of rectangles this chokehold made invalid
+					curr_sum -= (rel_streaks.back().second + 1)
+							* (rel_streaks.back().first - longest_consec[r][c]);
+					// add the number of kills this one got & kill this one itself
+					popped += rel_streaks.back().second + 1;
+					rel_streaks.pop_back();
+				}
+				rel_streaks.push_back({longest_consec[r][c], popped});
+				// add the submatrices of height 1
+				curr_sum += longest_consec[r][c];
+				total += curr_sum;
+			}
+		}
+		cout << total << '\n';
+	}
+}
+```
+
+</CPPSection>
+
+
+<JavaSection>
+
+```java
+import java.io.*;
+import java.util.*;
+
+public final class Rectangles {
+	public static void main(String[] args) throws IOException {
+		BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
+		int testNum = Integer.parseInt(read.readLine());
+		for (int t = 0; t < testNum; t++) {
+			StringTokenizer initial = new StringTokenizer(read.readLine());
+			int rowNum = Integer.parseInt(initial.nextToken());
+			int colNum = Integer.parseInt(initial.nextToken());
+			int[][] grid = new int[rowNum][];
+			for (int r = 0; r < rowNum; r++) {
+				grid[r] = Arrays.stream(read.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
+			}
+
+			int[][] longestConsec = new int[rowNum][colNum];
+			for (int r = 0; r < rowNum; r++) {
+				int curr = -1;
+				int streak = 0;
+				for (int c = colNum - 1; c >= 0; c--) {
+					if (grid[r][c] == curr) {
+						streak++;
+					} else {
+						curr = grid[r][c];
+						streak = 1;
+					}
+					longestConsec[r][c] = streak;
+				}
+			}
+
+			long total = 0;
+			for (int c = 0; c < colNum; c++) {
+				ArrayDeque<int[]> relStreaks = new ArrayDeque<>();
```

Comment:
Can u use a Stack instead? ArrayDeque is less known

Reply:
really?

Reply:
yeah, also you don't need the functionality of an arraydeque here, all the operations you use can be done with a stack, but this is up to u

Reply:
k cool